internal/querycoord/task_scheduler.go

```diff
@@ -608,53 +603,61 @@ func (scheduler *TaskScheduler) scheduleLoop() {
+		// after query coord restart, it will redo the triggerTask
+		// after recover, if child Task's state is TaskDone, will not be repeatedly executed
 		if err != nil {
+			log.Error("scheduleLoop: error when remove trigger task from etcd", zap.Int64("triggerTaskID", triggerTask.getTaskID()), zap.Error(err))
 			panic(err)
 		}
-		return nil
 	}
 
 	for {
 		var err error
 		select {
 		case <-scheduler.ctx.Done():
-			scheduler.stopActivateTaskLoopChan <- 1
 			return
 		case <-scheduler.triggerTaskQueue.Chan():
+			hasNotify := false
 			triggerTask = scheduler.triggerTaskQueue.popTask()
 			log.Debug("scheduleLoop: pop a triggerTask from triggerTaskQueue", zap.Int64("triggerTaskID", triggerTask.getTaskID()))
-			alreadyNotify := true
+			// the persistent task state is only taskUndo, taskDoing, taskDone
+			// taskUndo and taskDoing indicate that the assignment of internTasks for triggerTask was not completed
+			// taskDone said it had successfully completed the internTasks allocation for triggerTask
 			if triggerTask.getState() == taskUndo || triggerTask.getState() == taskDoing {
+				// if err == nil, the state of the triggerTask is already taskDone
+				// if err != nil, set the state of he triggerTask to taskFailed
 				err = scheduler.processTask(triggerTask)
 				if err != nil {
 					log.Debug("scheduleLoop: process triggerTask failed", zap.Int64("triggerTaskID", triggerTask.getTaskID()), zap.Error(err))
-					alreadyNotify = false
+					triggerTask.setState(taskFailed)
+				} else {
+					if triggerTask.msgType() == commonpb.MsgType_LoadCollection || triggerTask.msgType() == commonpb.MsgType_LoadCollection {
+						hasNotify = true
+						triggerTask.notify(nil)
+					}
 				}
 			}
-			if triggerTask.msgType() != commonpb.MsgType_LoadCollection && triggerTask.msgType() != commonpb.MsgType_LoadPartitions {
-				alreadyNotify = false
-			}
 
-			childTasks := triggerTask.getChildTask()
-			if len(childTasks) != 0 {
-				// process loadSegment before watchDmChannel, avoid delete not taking effect
-				highPriorityTasks, lowPriorityTasks := sortInternalTaskByPriority(childTasks, commonpb.MsgType_LoadSegments)
-				processInternalTaskFn(highPriorityTasks, triggerTask)
-				if triggerTask.getResultInfo().ErrorCode == commonpb.ErrorCode_Success {
-					processInternalTaskFn(lowPriorityTasks, triggerTask)
-				}
-				if triggerTask.getResultInfo().ErrorCode == commonpb.ErrorCode_Success {
-					err = updateSegmentInfoFromTask(scheduler.ctx, triggerTask, scheduler.meta)
-					if err != nil {
-						triggerTask.setResultInfo(err)
+			// int this step, triggerTask's state is either taskDone or taskFailed
+			if triggerTask.getState() == taskDone {
+				childTasks := triggerTask.getChildTask()
+				if len(childTasks) != 0 {
+					// process loadSegment before watchDmChannel, avoid delete not taking effect
+					highPriorityTasks, lowPriorityTasks := sortInternalTaskByPriority(childTasks, commonpb.MsgType_LoadSegments)
```

Comment:
The sortInternalTaskByPriority function name is confused, which is implementing the grouping function by MegType.